From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/pci: handle events for unused functions
Patch-mainline: v4.3-rc1
Git-commit: 515f022e8b59ce928fe5d82affb3e93ab53d12ed
References: bnc#946214, LTC#130628

Description:  s390/pci: fix handling of hotplug events
Symptom:      A pci function added to the system via hotplug
              might not be usable to Linux. In rare situations
              the kernel might crash.
Problem:      Linux does not handle all hotplug events correctly.
Solution:     Handle hotplug events correctly.
Reproduction: pci hotplug

Upstream-Description:

              s390/pci: handle events for unused functions

              Receiving error events for a pci function that's currently not in use
              will crash the kernel. For example the procedure for FW upgrades might
              include:
              * remove the function from Linux
              * apply FW upgrade
              * rescan for new functions

              Receiving an event during the FW upgrade will result in a use after free
              when printing the functions name. Just print "n/a" in such cases.

              Reviewed-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
              Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci_event.c |    6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

--- a/arch/s390/pci/pci_event.c
+++ b/arch/s390/pci/pci_event.c
@@ -46,15 +46,13 @@ struct zpci_ccdf_avail {
 static void __zpci_event_error(struct zpci_ccdf_err *ccdf)
 {
 	struct zpci_dev *zdev = get_zdev_by_fid(ccdf->fid);
+	struct pci_dev *pdev = zdev ? zdev->pdev : NULL;
 
 	zpci_err("error CCDF:\n");
 	zpci_err_hex(ccdf, sizeof(*ccdf));
 
-	if (!zdev)
-		return;
-
 	pr_err("%s: Event 0x%x reports an error for PCI function 0x%x\n",
-	       pci_name(zdev->pdev), ccdf->pec, ccdf->fid);
+	       pdev ? pci_name(pdev) : "n/a", ccdf->pec, ccdf->fid);
 }
 
 void zpci_event_error(void *data)
